include/wil/safearrays.h

```diff
@@ -0,0 +1,897 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+#ifndef __WIL_SAFEARRAYS_INCLUDED
+#define __WIL_SAFEARRAYS_INCLUDED
+
+#ifdef _KERNEL_MODE
+#error This header is not supported in kernel-mode.
+#endif
+
+#include <new.h> // new(std::nothrow)
+#include "wistd_type_traits.h"
+#include "resource.h" // unique_hkey
+
+namespace wil
+{
+#if defined( _OLEAUTO_H_ ) && !defined(__WIL_SAFEARRAY_)
+#define __WIL_SAFEARRAY_
+    /// @cond
+    namespace details
+    {
+        template<typename T>
+        struct var_traits{};

```

Comment:
The name `var_traits` is a bit generic and could apply to a number of scenarios and would probably benefit from a different name

Reply:
I was originally going to re-use these for my VARIANT helper classes that I was going to do next, and in that case, they have more work to do and `var_traits<>` is absolutely a great name for them.  Since I'm not sure I'm going to do the VARIANT classes, I'll rename them and revisit the issue if/when I do.

I'll rename to `infer_safearray_type_traits`.